lib/compress.js

```diff
@@ -1344,37 +1354,38 @@ merge(Compressor.prototype, {
         def(AST_UnaryPrefix, function() {
             return this.operator == "void";
         });
-        def(AST_Binary, function(pure_getters) {
+        def(AST_Binary, function(compressor) {
             switch (this.operator) {
               case "&&":
-                return this.left._throw_on_access(pure_getters);
+                return this.left._dot_throw(compressor);
               case "||":
-                return this.left._throw_on_access(pure_getters)
-                    && this.right._throw_on_access(pure_getters);
+                return this.left._dot_throw(compressor)
+                    && this.right._dot_throw(compressor);
               default:
                 return false;
             }
         })
-        def(AST_Assign, function(pure_getters) {
+        def(AST_Assign, function(compressor) {
             return this.operator == "="
-                && this.right._throw_on_access(pure_getters);
+                && this.right._dot_throw(compressor);
         })
-        def(AST_Conditional, function(pure_getters) {
-            return this.consequent._throw_on_access(pure_getters)
-                || this.alternative._throw_on_access(pure_getters);
+        def(AST_Conditional, function(compressor) {
+            return this.consequent._dot_throw(compressor)
+                || this.alternative._dot_throw(compressor);
         })
-        def(AST_Sequence, function(pure_getters) {
-            return this.expressions[this.expressions.length - 1]._throw_on_access(pure_getters);
+        def(AST_Sequence, function(compressor) {
+            return this.expressions[this.expressions.length - 1]._dot_throw(compressor);
         });
-        def(AST_SymbolRef, function(pure_getters) {
+        def(AST_SymbolRef, function(compressor) {
             if (this.is_undefined) return true;
-            if (!is_strict(pure_getters)) return false;
+            if (!is_strict(compressor)) return false;
+            if (is_undeclared_ref(this) && this.is_declared(compressor)) return false;
             if (this.is_immutable()) return false;
             var fixed = this.fixed_value();
-            return !fixed || fixed._throw_on_access(pure_getters);
+            return !fixed || fixed._dot_throw(compressor);
         });
     })(function(node, func) {
-        node.DEFMETHOD("_throw_on_access", func);
+        node.DEFMETHOD("_dot_throw", func);
```

Comment:
I thought `_throw_on_access` was a suitable name. Presumably `_dot_throw` throws for both `AST_Dot` and `AST_Sub` invalid access?

BTW, I like the fact that the parameter was cleaned up and changed from `pure_getters` to `compressor`.